snorkel/slicing/utils.py

```diff
@@ -0,0 +1,202 @@
+        dataloader.task_to_label_dict[pred_task_name] = pred_task_name
+    return dataloader
+
+
+def convert_to_slice_tasks(base_task: Task, slice_names: List[str]):
+    """Adds slice labels to dataloader and creates new slice tasks (including base slice)
+
+    Each slice will get two slice-specific heads:
+    - an indicator head that learns to identify when DataPoints are in that slice
+    - a predictor head that is trained on only members of that slice
+
+    The base task's head is replaced by a master head that makes predictions based on
+    a combination of the predictor heads' predictions that are weighted by the
+    indicator heads' predictions.
+
+    NOTE: The current implementation pollutes the module_pool---the indicator's task's
+    module_pool includes predictor modules and vice versa since both are modified in
+    place. This does not affect the result because the task flows dictate which modules
+    get used, and those do not include the extra modules. An alternative would be to
+    make separate copies of the module pool for each, but that wastes time and memory
+    extra copies of (potentially very large) modules that will be merged in a moment
+    away in the model since they have the same name. We leave resolution of this issue
+    for a future release.
+    """
+    if "base" not in slice_names:
+        slice_names = slice_names + ["base"]
+
+    tasks = []
+
+    # NOTE: We assume here that the last operation uses the head module
+    # Identify base task head module
+    head_module_op = base_task.task_flow[-1]
+    head_module = base_task.module_pool[head_module_op.module_name]
+
+    if isinstance(head_module, nn.DataParallel):
+        head_module = head_module.module
+
+    neck_size = head_module.in_features
+    base_task_cardinality = head_module.out_features
+
+    # Remove the slice-unaware head module from module pool and task flow
+    del base_task.module_pool[head_module_op.module_name]
+    base_task.task_flow = base_task.task_flow[:-1]
+
+    # Create slice indicator tasks
+    for slice_name in slice_names:
+
+        ind_task_name = f"{base_task.name}_slice:{slice_name}_ind"
+
+        ind_head_module_name = f"{ind_task_name}_ind_head"
+        ind_head_module = nn.Linear(neck_size, 2)
+
+        # Create module_pool
+        ind_module_pool = base_task.module_pool
+        ind_module_pool[ind_head_module_name] = ind_head_module
+
+        # Create task_flow
+        ind_head_op = Operation(
+            module_name=ind_head_module_name, inputs=head_module_op.inputs
+        )
+        ind_task_flow = base_task.task_flow + [ind_head_op]
+
+        # Create ind task
+        ind_task = Task(
+            name=ind_task_name,
+            module_pool=ind_module_pool,
+            task_flow=ind_task_flow,
+            loss_func=partial(ce_loss, ind_head_op.name),
+            output_func=partial(softmax, ind_head_op.name),
+            scorer=Scorer(metrics=["accuracy"]),
```

Comment:
let's add `f1` by default here—usually indicators have class balance issues

Reply:
Yeah, I was hesitant because the metrics dict quickly balloons in size, but from a practical standpoint, it is nice to have it. Adding.

Reply:
if we want to reduce size, i think f1 is preferable to accuracy, fwiw